dev/ToastNotifications/ToastNotificationManager.cpp

```diff
@@ -0,0 +1,135 @@
+﻿#include "pch.h"
+#include "ToastNotificationManager.h"
+#include "Microsoft.Windows.ToastNotifications.ToastNotificationManager.g.cpp"
+#include <winrt/Windows.ApplicationModel.Core.h>
+#include <winrt/Windows.Storage.h>
+#include <winrt/base.h>
+#include "ToastActivationCallback.h"
+#include "externs.h"
+#include "ToastNotificationUtility.h"
+#include <frameworkudk/pushnotifications.h>
+
+static winrt::event<winrt::Windows::Foundation::EventHandler<winrt::Microsoft::Windows::ToastNotifications::ToastActivatedEventArgs>> g_toastHandlers;
+
+static int g_toastHandlerCount = 0;
+
+winrt::event<winrt::Windows::Foundation::EventHandler<winrt::Microsoft::Windows::ToastNotifications::ToastActivatedEventArgs>>& GetToastHandlers()
+{
+    return g_toastHandlers;
+}
+
+int& GetToastHandleCount()
+{
+    return g_toastHandlerCount;
+}
+
+namespace winrt
+{
+    using namespace winrt::Windows::UI;
+    using namespace winrt::Windows::Foundation;
+    using namespace Windows::ApplicationModel::Core;
+}
+
+namespace winrt::Microsoft::Windows::ToastNotifications::implementation
+{
+    static wil::unique_com_class_object_cookie s_toastcomActivatorRegistration;
+
+    winrt::Microsoft::Windows::ToastNotifications::ToastNotificationManager ToastNotificationManager::Default()
+    {
+        throw hresult_not_implemented();
+    }
+
+    void ToastNotificationManager::RegisterActivator(winrt::Microsoft::Windows::ToastNotifications::ToastActivationInfo const&  details)
+    {
+        THROW_HR_IF_MSG(E_INVALIDARG, s_toastcomActivatorRegistration, "ComActivator already registered.");
+
+        GUID comActivatorGuid = GUID_NULL;
+        if (!AppModel::Identity::IsPackagedProcess())
+        {
+            wil::unique_cotaskmem_string processName;
+            THROW_IF_FAILED(GetCurrentProcessPath(processName));
+
+            std::wstring appIdentifier{ RetrieveToastGuid() };
+            THROW_IF_FAILED(PushNotifications_RegisterFullTrustApplication(appIdentifier.c_str(), GUID_NULL));
+
+            std::wstring storedComActivatorString{ RetrieveComActivatorGuid(appIdentifier) };
+
+            if (storedComActivatorString.empty())
```

Comment:
We should move this empty string check to RetrieveComActivatorGUID. It should create a new GUID if it hasn't created one already

Reply:
ie. It should mimic RetreiveToastGuid above

Reply:
I believe I used the stored com activator as the condition for whether or not we have registered assets/comserver. If we move the com activator guid creation inside RetrieveComActivatorGuid (which makes more sense), we don't know if have registered assets/comserver before. We could reregister everything but I thought that would be weird. Another option is to return a std::pair<winrt::guid, bool> for whether or not we were created yet. What do you think?

Reply:
RetrieveCOMActivatorGuid can take in the Assets along with AppIdentifier and replace the assets in registry. I think its ok to replace existing Assets if they exist already: no harm. The app just decided to change its branding(Display name and icon) which is a real scenario right?

Reply:
And you can rename the function to RegisterComAssetsAndGuid. It takes in appid and assets. Figures out  a new com activator guid if not present already and re-registers everything.